earn/src/components/borrow/modal/RepayModal.tsx

```diff
@@ -1,109 +1,231 @@
-    enabled:
-      !!userAddress && repayAmount.gt(0) && repayAmount.lte(liabilityAmountBig) && repayAmount.lte(collateralAmount),
+  // MARK: Fetching and preparing data that's necessary to figure out button state ------------------------------------
+  // --> How much `repayToken` does user have in their wallet?
+  const { data: repayTokenBalance } = useBalance({
+    address: userAddress,
     chainId: activeChain.id,
+    token: repayToken.address,
+    watch: false,
   });
-  const removeCollateralUpdatedRequest = useMemo(() => {
-    if (removeCollateralConfig.request) {
-      return {
-        ...removeCollateralConfig.request,
-        gasLimit: removeCollateralConfig.request.gasLimit.mul(GAS_ESTIMATE_WIGGLE_ROOM).div(100),
-      };
+  const numericRepayTokenBalance = parseFloat(repayTokenBalance?.formatted ?? '0') || 0;
+  // --> Are they trying to repay more than their total liability?
+  const numericRepayAmount = parseFloat(repayAmount) || 0;
+  const existingLiability = marginAccount.liabilities[lender === marginAccount.lender0 ? 'amount0' : 'amount1'];
+  // --> Other...
+  const bigRepayAmount = ethers.utils.parseUnits(repayAmount === '' ? '0' : repayAmount, repayToken.decimals);
+
+  // MARK: Determining button state -----------------------------------------------------------------------------------
+  let confirmButtonState: ConfirmButtonState;
+
+  if (repayAmount === '') {
+    confirmButtonState = ConfirmButtonState.DISABLED;
+  } else if (numericRepayAmount > numericRepayTokenBalance) {
+    confirmButtonState = ConfirmButtonState.INSUFFICIENT_FUNDS;
+  } else if (numericRepayAmount > existingLiability) {
+    confirmButtonState = ConfirmButtonState.REPAYING_TOO_MUCH;
+  } else if (permitDomain !== null) {
+    if (!permitData) {
+      confirmButtonState = ConfirmButtonState.PERMIT_ASSET;
+    } else {
+      confirmButtonState = ConfirmButtonState.READY_VIA_PERMIT;
+    }
+  } else if (permitDomain === null && routerAllowance) {
+    if (routerAllowance.lt(bigRepayAmount)) {
+      confirmButtonState = ConfirmButtonState.APPROVE_ASSET;
+    } else {
+      confirmButtonState = ConfirmButtonState.READY_VIA_APPROVE;
     }
-    return undefined;
-  }, [removeCollateralConfig.request]);
+  } else if (isPending) {
+    confirmButtonState = ConfirmButtonState.PENDING;
+  } else {
+    console.error('Unexpected confirm button state!');
+    confirmButtonState = ConfirmButtonState.DISABLED;
+  }
+
+  // MARK: Prepare contract write for approval flow -------------------------------------------------------------------
+  const { config: repayWithApprovalConfig, refetch: refetchRepayWithApprovalConfig } = usePrepareContractWrite({
+    address: ALOE_II_ROUTER_ADDRESS,
+    abi: RouterABI,
+    functionName: 'repay(address,uint256,address)',
+    args: [lender, bigRepayAmount, marginAccount.address],
+    chainId: activeChain.id,
+    enabled: confirmButtonState === ConfirmButtonState.READY_VIA_APPROVE,
```

Comment:
In the earn interest modal, we define these hooks above the place where we determine button state. Here I put them below, so that the `enabled` property is simpler